tulip/abstract/prop2partition.py

```diff
@@ -30,83 +30,305 @@
 # OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 # SUCH DAMAGE.
 #
-"""
-Proposition preserving partition module.
-"""
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import print_function
-
+"""Proposition-preserving partition module."""
+import copy
 import logging
-logger = logging.getLogger(__name__)
-
+import typing as _ty
 import warnings
-import copy
 
 import numpy as np
-from scipy import sparse as sp
 import polytope as pc
-from polytope.plot import plot_partition
-
-from tulip import transys as trs
+import polytope.plot as _pplot
+import scipy.sparse as sp
 
-# inline imports:
-#
-# from tulip.graphics import newax
+import tulip.graphics as _graphics
+_mpl = _graphics._mpl
+import tulip.transys as trs
 
 
 __all__ = [
-    'prop2part', 'part2convex',
-    'pwa_partition', 'add_grid',
-    'PropPreservingPartition', 'PPP', 'ppp2ts']
+    'PropPreservingPartition',
+    'PPP',
+    'prop2part',
+    'part2convex',
+    'pwa_partition',
+    'add_grid',
+    'ppp2ts']
 
 
 _hl = 40 * '-'
+_logger = logging.getLogger(__name__)
+Polytope = (
+    pc.Polytope |
+    pc.Region)
+
+
+class PropPreservingPartition(
+        pc.MetricPartition):
+    """Partition class with following fields:
+
+    Attributes:
 
-def prop2part(state_space, cont_props_dict):
-    """Main function that takes a domain (state_space) and a list of
-    propositions (cont_props), and returns a proposition preserving
-    partition of the state space.
+    - `domain`: the domain we want to partition
 
-    See Also
+    - `regions`: `Region`s of
+      proposition-preserving partition
+
+    - `adj`: sparse matrix showing which
+      regions are adjacent.
+      The order of `Region`s is the
+      same as in the `list` `regions`.
+
+      - type: `scipy` `lil` `sparse`
+
+    - `prop_regions`: map from atomic
+      proposition symbols to continuous subsets
+
+
+    Relevant
     ========
-    L{PropPreservingPartition},
-    C{polytope.Polytope}
+    `prop2part`
+    """
+    def __init__(
+            self,
+            domain:
+                Polytope |
+                None=None,
+            regions:
+                list[Polytope] |
+                None=None,
+            adj:
+                sp.lil_array |
+                None=None,
+            prop_regions:
+                dict[..., Polytope] |
+                None=None,
+            check:
+                bool=True):
+        if regions is None:
+            regions = list()
+        # super().__init__(adj)
```

Comment:
This comment can be deleted because `super().__init__()` is called below and the original comment is from commit 3d3383beecce942a6499e3ee0f2d8e953e1ec3b2, not relevant anymore.